Feat: Add cdss resources

eastgerm opened the pull request:
## 요약
- CDSS 신규 상품 통합 개발
- 관련 docs, examples 추가

## 내용
### Resource
- ncloud_cdss_cluster
- ncloud_cdss_config_group
### Datasource
- ncloud_cdss_cluster
- ncloud_cdss_config_group
- ncloud_cdss_os_product
- ncloud_cdss_node_product
- ncloud_cdss_kafka_version

## 비고
- public endpoint를 위한 loadbalancer 연동 시나리오에 대한 개발은 내부 로직 변경이 필요하여 우선 제외하였습니다. 추후 업데이트 하는 방식으로 기능 추가 해보도록 하겠습니다.
- 코드 리뷰 및 test code 관련 안내주시면 감사하겠습니다.

## 테스트 환경
- 환경: GOV 베타
- go-sdk: https://github.com/eastgerm/ncloud-sdk-go-v2/tree/vcdss_support_config_group
### examples/vcdss/main.tf 실행(terraform plan) 결과
> https://github.com/eastgerm/terraform-provider-ncloud/blob/feature/vcdss/examples/vcdss/main.tf
```
user@AL02019676  ~/go/src/github.com/eastgerm/terraform-provider-ncloud/examples/vcdss   feature/vcdss  terraform plan
data.ncloud_cdss_cluster.cluster: Reading...
data.ncloud_cdss_kafka_version.kafka-version: Reading...
data.ncloud_cdss_os_product.os_sample: Reading...
data.ncloud_cdss_kafka_version.kafka-version: Read complete after 0s [id=2823006]
data.ncloud_cdss_cluster.cluster: Read complete after 0s [id=2683820]
data.ncloud_cdss_os_product.os_sample: Read complete after 0s [id=SW.VCDSS.OS.LNX64.CNTOS.0708.B050]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create
 <= read (data resources)

Terraform will perform the following actions:

  # data.ncloud_cdss_node_product.node_sample will be read during apply
  # (config refers to values not yet known)
 <= data "ncloud_cdss_node_product" "node_sample" {
      + cpu_count       = (known after apply)
      + id              = (known after apply)
      + memory_size     = (known after apply)
      + os_product_code = "SW.VCDSS.OS.LNX64.CNTOS.0708.B050"
      + product_type    = (known after apply)
      + subnet_no       = (known after apply)

      + filter {
          + name   = "cpu_count"
          + values = [
              + "2",
            ]
        }
      + filter {
          + name   = "memory_size"
          + values = [
              + "8GB",
            ]
        }
      + filter {
          + name   = "product_type"
          + values = [
              + "STAND",
            ]
        }
    }

  # ncloud_cdss_cluster.cluster-1 will be created
  + resource "ncloud_cdss_cluster" "cluster-1" {
      + config_group_no    = (known after apply)
      + id                 = (known after apply)
      + kafka_version_code = "2823006"
      + name               = "from-tf-cdss"
      + os_product_code    = "SW.VCDSS.OS.LNX64.CNTOS.0708.B050"
      + vpc_no             = (known after apply)

      + broker_nodes {
          + node_count        = "3"
          + node_product_code = (known after apply)
          + storage_size      = "100"
          + subnet_no         = (known after apply)
        }

      + cmak {
          + user_name     = "terraform"
          + user_password = (sensitive value)
        }

      + manager_node {
          + node_product_code = (known after apply)
          + subnet_no         = (known after apply)
        }
    }

  # ncloud_cdss_cluster.cluster-2 will be created
  + resource "ncloud_cdss_cluster" "cluster-2" {
      + config_group_no    = "26"
      + id                 = (known after apply)
      + kafka_version_code = "2823006"
      + name               = "asdfasdf"
      + os_product_code    = "SW.VCDSS.OS.LNX64.CNTOS.0708.B050"
      + vpc_no             = "3438"

      + broker_nodes {
          + node_count        = "3"
          + node_product_code = "SVR.VCDSS.STAND.C002.M008.NET.SSD.B050.G002"
          + storage_size      = "100"
          + subnet_no         = "4853"
        }

      + cmak {
          + user_name     = "string"
          + user_password = (sensitive value)
        }

      + manager_node {
          + node_product_code = "SVR.VCDSS.STAND.C002.M008.NET.SSD.B050.G002"
          + subnet_no         = "5573"
        }
    }

  # ncloud_cdss_config_group.config-group will be created
  + resource "ncloud_cdss_config_group" "config-group" {
      + description        = "test"
      + id                 = (known after apply)
      + kafka_version_code = "2823006"
      + name               = "from-tf-config"
    }

  # ncloud_subnet.private-subnet will be created
  + resource "ncloud_subnet" "private-subnet" {
      + id             = (known after apply)
      + name           = "from-tf-private"
      + network_acl_no = (known after apply)
      + subnet         = "10.234.100.0/24"
      + subnet_no      = (known after apply)
      + subnet_type    = "PRIVATE"
      + usage_type     = "GEN"
      + vpc_no         = (known after apply)
      + zone           = "KR-2"
    }

  # ncloud_subnet.public-subnet will be created
  + resource "ncloud_subnet" "public-subnet" {
      + id             = (known after apply)
      + name           = "from-tf-public"
      + network_acl_no = (known after apply)
      + subnet         = "10.234.0.0/24"
      + subnet_no      = (known after apply)
      + subnet_type    = "PUBLIC"
      + usage_type     = "GEN"
      + vpc_no         = (known after apply)
      + zone           = "KR-2"
    }

  # ncloud_vpc.vpc will be created
  + resource "ncloud_vpc" "vpc" {
      + default_access_control_group_no = (known after apply)
      + default_network_acl_no          = (known after apply)
      + default_private_route_table_no  = (known after apply)
      + default_public_route_table_no   = (known after apply)
      + id                              = (known after apply)
      + ipv4_cidr_block                 = "10.234.0.0/16"
      + name                            = "from-tf-vpc"
      + vpc_no                          = (known after apply)
    }

Plan: 6 to add, 0 to change, 0 to destroy.

────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
```

Commit:
feature: support vcdss

eastgerm committed:
Update resource, datasource about CDSS

eastgerm committed:
Update /docs about CDSS

wonchulee commented on examples/vcdss/variables.tf:
```diff
@@ -0,0 +1,15 @@
+
+variable secret_key {
+  default = "oRdQ4fRhB2DspQlatTjoONzP9myqOWito13De0ef"
+}
```
임의의 값을 넣어주신 것일 것 같지만, 그렇더라도 노출되는 키값은 의미없는 것으로 넣어주시면 좋겠습니다. ex) "YOUR_ACCESS_KEY"

minosmlee replied:
아래와 같이 아예 환경변수나 콘솔을 통해입력받도록 해주시는게 제일 좋습니다.
```
variable access_key {}
variable secret_key {}
```

wonchulee commented on ncloud/data_source_ncloud_cdss_config_group.go:
```diff
@@ -0,0 +1,58 @@
+		return diag.FromErr(NotSupportClassic("dataSource `ncloud_cdss_config_group`"))
+	}
+
+	configGroup, err := getCDSSConfigGroup(ctx, config, *StringPtrOrNil(d.GetOk("kafka_version_code")), *StringPtrOrNil(d.GetOk("id")))
```
의미상으로 `StringPtrOrNil()` 은 String Ptr 타입의 값 또는 Nil 둘 중 하나를 리턴한다는 네이밍을 갖는 것으로 보입니다. 문맥상 nil 값이 될 수 없다고 하더라도 nil이 될수 있다는 네이밍의 함수를 역참조하는 것`*StringPtrOrNil()`은 조금 이상하게 느껴질 수 있을 것 같습니다.
예를들면 `UnwrapStrPtr()` 과 같이 명시적으로 Nil이 아닌 값만을 사용한다는 의미를 담고, GetOk()의 결과가 ok = false 인 경우 패닉을 발생시키는 형태의 helper 함수를 사용하는 것이 좋겠습니다.

eastgerm replied:
@wonchulee 좋은 의견 감사합니다. 
`UnwrapStrPtr()`이란 함수를 만들어 사용하면 좋을듯한데, 다른 상품들의 코드를 봤을때 이미 `*StringPtrOrNil()`, `*getInt32FromString()`를 많은 곳에서 사용하고 있는거로 보여집니다. (다른 상품 코드에서 참고한 부분이기도 하구요) 

`UnwrapStrPtr()`과 같은 함수를 사용한다면 이게 전체적인 통일성을 해치는게 되지는 않을런지요? 의견 주시는대로 적용해보겠습니다.

wonchulee replied:
`StringPtrOrNil()` 과 `getInt32FromString()`를 역참조 해서 사용하지는 않았었는데, go sdk의 Response 인자가 String pointer를 주로 사용하고 있었지만 여기에서는 String 타입이 사용되어서 어색하게 보일 수 있는 부분이었네요.
해당 부분은 동작 상 이슈가 되지 않는 부분이니, 반영 이후에 제가 별도로 해당 함수를 사용하는 코드 전반을 정리하도록 하겠습니다.

wonchulee commented on ncloud/data_source_ncloud_cdss_node_product.go:
```diff
@@ -0,0 +1,115 @@
+
+	reqParams := vcdss.NodeProduct{
+		SoftwareProductCode: *StringPtrOrNil(d.GetOk("os_product_code")),
+		SubnetNo:            *getInt32FromString(d.GetOk("subnet_no")),
```
getInt32FromString() 역참조 또한 위의 StringPtrOrNil()의 코멘트를 참고부탁드립니다.

wonchulee commented on ncloud/data_source_ncloud_cdss_node_product.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+func parseMemorySize(memorySize string) string {
+	num, _ := strconv.Atoi(memorySize)
```
사소한 부분이지만, `Atoi()`에서 err가 반환되는 경우 에러를 함께 전달하여 getCDSSNodeProducts()에서의 예외처리를 추가해주시면 좋을 것 같습니다.

wonchulee commented on ncloud/resource_ncloud_cdss_cluster.go:
```diff
@@ -0,0 +1,405 @@
+	StatusError    = "error"
+	StatusReturn   = "return"
+	StatusNull     = "null"
+)
```
현재 모든 리소스가 ncloud package에 포함되어 있어, cdss cluster 에서만 사용될 목적의 Status값이라면 CDSSCluster prefix를 포함시켜서 네이밍 해주시면 좋을 것 같습니다!

eastgerm committed:
Change TypeSet to TypeList

wonchulee commented on examples/vcdss/versions.tf:
```diff
@@ -0,0 +1,8 @@
+      source = "terraform.local/local/ncloud"
+      version = "1.0.0"
+    }
+  }
```
버전 정보는 추후에 반영하기 전에 아래와 유사하게 수정되면 좋을 것 같습니다.
```
terraform {
  required_version = ">= 0.13"
  required_providers {
    ncloud = {
      source = "navercloudplatform/ncloud"
    }
  }
}
```

wonchulee commented:
안녕하세요, 사소한 마이너한 내용의 코멘트들 남겼습니다 참고부탁드려요. 베타 환경에서 동작을 확인해보는 중인데, 혹시 필요한 내용이 있다면 추가적으로 코멘트 남기도록 하겠습니다. 감사합니다.

minosmlee commented on docs/data-sources/cdss_node_product.md:
```diff
@@ -0,0 +1,48 @@
+    values = ["2"]
+  }
+
+  filter {
+    name   = "memory_size"
+    values = ["8GB"]
+  }
```
memory 관련 필터가 정상 동작하지 않습니다. 해당 필터 제외하고 output을 찍어보았을때 결과는 아래와같습니다.
``` hcl
node_product = {
  "cpu_count" = "2"
  "filter" = toset(null) /* of object */
  "id" = "SVR.VCDSS.STAND.C002.M008.NET.SSD.B050.G002"
  "memory_size" = "0GB"  <- 정상적으로 찍히지 않는 상황, 해당 값으로 필터링 동작하지 않음
  "os_product_code" = "SW.VCDSS.OS.LNX64.CNTOS.0708.B050"
  "product_type" = "STAND"
  "subnet_no" = "59354"
}
```

minosmlee replied:
참고로 민간 리얼에서 테스트했습니다.

eastgerm replied:
@minosmlee, 네 해당 부분은 기존 api에서 변경이 필요했어서 현재 **공공 베타**에서 동작 확인이 가능합니다

wonchulee replied:
공공 베타에서도 동일한 현상이 있는 것 같습니다. 혹시 문제가 있는지 한 번 확인부탁드려요

minosmlee commented on ncloud/resource_ncloud_cdss_cluster.go:
```diff
@@ -0,0 +1,406 @@
+			"endpoints": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
```
cluster 생성 후 변경 사항 없이 terraform apply 수행 시 endpoint 속성 때문에 replace 발생합니다.
```
Optional: true -> Computed: true 변경
ForceNew: true -> 제거
```

endpoints 하위 속성 모두
```
Optional: true  -> Computed: true 변경
```

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,130 @@
+  cmak {
+    user_name = "terraform"
+    user_password = "test123!@#"
+  }
```
예시에 패스워드를 직접 입력하도록 가이드하는것은 좋지 않습니다.
아래와같이 가이드를 변경하면 좋겠습니다.

``` hcl
variable "cmak_user_password" {
  description = "CDSS cluster CMAK user password"
  type        = string
  sensitive   = true
}

resource "ncloud_cdss_cluster" "cluster" {
  cmak {
    user_name     = "terraform"
    user_password = var.cmak_user_password
  }
}

```

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,130 @@
+
+## Attribute Reference
+In addition to all arguments above, the following attributes are exported
+
```
CDSS 클러스터를 생성한 뒤 다른 테라폼 리소스와 원활한 연동을 위해서는 필히 아래의 어트리뷰트가 output으로 노출되어야할거같습니다. 만약 지금 당장 지원이 안된다면 선 릴리즈 후 향후 지원이 되어야합니다.
- 매니저 노드 ACG ID
- 브로커 노드 ACG ID
- 매니저 노드 인스턴스 정보 (Name, ID)
- 브로커 노드 인스턴스 정보 (Name, ID)

minosmlee replied:
output으로 timeout이 출력됩니다. 이부분은 노출이 되는것이 맞을까요? 아니라면 출력이 되지 않거나, 가이드에 포함이 되어야합니다.

eastgerm replied:
> CDSS 클러스터를 생성한 뒤 다른 테라폼 리소스와 원활한 연동을 위해서는 필히 아래의 어트리뷰트가 output으로 노출되어야할거같습니다. 만약 지금 당장 지원이 안된다면 선 릴리즈 후 향후 지원이 되어야합니다.
> 
> * 매니저 노드 ACG ID
> * 브로커 노드 ACG ID
> * 매니저 노드 인스턴스 정보 (Name, ID)
> * 브로커 노드 인스턴스 정보 (Name, ID)

@minosmlee 네 해당 내용을 지원하려면 open api의 변경이 이루어져야하는데 12월 출시 일정에 맞춰 변경을 적용하기엔 타이트할듯 합니다. 이부분은 릴리즈 후 향후 지원 하도록 하겠습니다. 감사합니다

eastgerm replied:
> output으로 timeout이 출력됩니다. 이부분은 노출이 되는것이 맞을까요? 아니라면 출력이 되지 않거나, 가이드에 포함이 되어야합니다.
```golang
return &schema.Resource{
		CreateContext: resourceNcloudCDSSClusterCreate,
		ReadContext:   resourceNcloudCDSSClusterRead,
		DeleteContext: resourceNcloudCDSSClusterDelete,
		Importer: &schema.ResourceImporter{
			StateContext: schema.ImportStatePassthroughContext,
		},
		Timeouts: &schema.ResourceTimeout{
			Create: schema.DefaultTimeout(DefaultCreateTimeout),
			Update: schema.DefaultTimeout(DefaultCreateTimeout),
			Delete: schema.DefaultTimeout(DefaultCreateTimeout),
		},
```
@minosmlee 의도한 부분은 아니지만 위의 `Timeouts` 설정으로 인한 output으로 보이는데 가이드에 추가하도록 하겠습니다.

minosmlee replied:
Timeout은 무시해도 될 것같습니다.

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,130 @@
+    * `public_endpoint_broker_tls_node_listener_port_list` - List of listener port for public endpoint of broker nodes (TLS).
+    * `local_dns_list` - Editing details of the hosts file (Private IP hostname format).
+    * `local_dns_tls_list` - Editing details of the hosts file (Private IP hostname format) (TLS).
+    * `zookeeper_list` - List of ZooKeeper nodes (Port 2181).
```
콘솔에 노출되는 값과 최대한 이름과 구조를 일치시켰으면 합니다. (가능하다면)
``` hcl
endpoints = {
  broker_node = {
    plaintext = "10.0.1.6:9092,10.0.1.10:9092,10.0.1.11:9092"
    tls       = "cdss-tf-b-22uw:9093,cdss-tf-b-22ux:9093,cdss-tf-b-22uy:9093"
    public_endpoint_plaintext_listener_port = <<-EOT
      cdss-tf-b-22uw 9100
      cdss-tf-b-22ux 9101
      cdss-tf-b-22uy 9102
      
      EOT
    public_endpoint_tls_listener_port = <<-EOT
      cdss-tf-b-22uw 9300
      cdss-tf-b-22ux 9301
      cdss-tf-b-22uy 9302
      
      EOT
  }
  zookeeper = "10.0.1.6:2181,10.0.1.10:2181,10.0.1.11:2181"
  hosts = {
    private_endpoint_tls = <<-EOT
      10.0.1.6 cdss-tf-b-22uw
      10.0.1.10 cdss-tf-b-22ux
      10.0.1.11 cdss-tf-b-22uy

      EOT
  }
}
```


그런데, <<-EOT / EOT 사이의 구문은 저형태로 노출되어야하는걸까요? 그렇지 않다면 (예를들어 그대로 복사해서 사용해야하는 경우가 아니라면) list 형태로 감싸서 노출되는것이 좋을것같습니다.
``` hcl
public_endpoint_tls_listener_port = [
  "cdss-tf-b-22uw 9300",
  "cdss-tf-b-22ux 9301",
  "cdss-tf-b-22uy 9302",
]
```

eastgerm replied:
@minosmlee 음 최대한 위 구조로 진행해보려 했으나 뎁스가 2이상 되었을 때 값을 set하는 방법을 알아내지 못해서.. 이름만 콘솔과 일치하게 변경하겠습니다

<<-EOT / EOT 사이의 구문은 콘솔에서도 그대로 복사해서 사용하라는 의도로 제공하고 있기 때문에 그대로 유지해도 괜찮을 듯 합니다

eastgerm replied:
## 이름 변경
```tf
"endpoints" = tolist([
    {
      "hosts_private_endpoint_tls" = <<-EOT
         10.235.100.6 from-tf-cdss-b-eup
         10.235.100.7 from-tf-cdss-b-euq
         10.235.100.8 from-tf-cdss-b-eur
      
         EOT
      "hosts_public_endpoint_tls" = ""
      "plaintext" = "10.235.100.6:9092,10.235.100.7:9092,10.235.100.8:9092"
      "public_endpoint_broker_tls_node_listener_port_list" = ""
      "public_endpoint_plaintext" = ""
      "public_endpoint_plaintext_listener_port" = <<-EOT
         from-tf-cdss-b-eup 9100
         from-tf-cdss-b-euq 9101
         from-tf-cdss-b-eur 9102
      
         EOT
      "public_endpoint_tls" = ""
      "public_endpoint_tls_listener_port" = <<-EOT
         from-tf-cdss-b-eup 9300
         from-tf-cdss-b-euq 9301
         from-tf-cdss-b-eur 9302
      
         EOT
      "tls" = "from-tf-cdss-b-eup:9093,from-tf-cdss-b-euq:9093,from-tf-cdss-b-eur:9093"
      "zookeeper" = "10.235.100.6:2181,10.235.100.7:2181,10.235.100.8:2181"
    },
  ])

```

wonchulee replied:
@minosmlee 님이 의견주신대로 이름을 변경해주셨네요! 위 의견과 상반된 내용이긴한데.. 개인적으로는 `plaintext, tls` 대신 기존 네이밍이 사용자 입장에서 조금 더 직관적으로 이해할 수 있는 이름일 것 같아, 기존 이름을 유지하는 것도 좋겠습니다. 하지만 이미 작업 해주셨기 때문에 이 의견은 참고로만 봐주셔도 괜찮을 것 같습니다.

wonchulee replied:
https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248/commits/e52b03499ba6d9a904a899cc577c756289f85ff9 에서 List 타입으로 변경

minosmlee commented on examples/vcdss/.terraform.lock.hcl:
```diff
@@ -0,0 +1,10 @@
+  hashes = [
+    "h1:qX2F7NTGWZn+NHAxY+3NSiEBSa7g6ybmRFtrWvZvAcA=",
+  ]
+}
```
.terraform.lock.hcl 은 제외해주세요~

minosmlee commented on docs/data-sources/cdss_cluster.md:
```diff
@@ -0,0 +1,70 @@
+## Argument Reference
+The following arguments are supported
+
+* `id` - (Required) Cluster uuid.
```
datasource의 용도는 name등의 속성을 이용하여 ID를 찾기 위한 용도가 메인입니다.
Filter 기능을 추가해주셔야할것같습니다.
example 역시 ID로 찾기보다 filter(name)을 이용하여 ID를 찾는 형태로 변경하는것이 좋을것같습니다.

또한 cluster_uuid == id 일 경우 cluster_uuid와 id를 동시에 같은 값으로 노출시키면 좋을것같습니다.
server의 예에도 id와 instanceNo 를 동시에 노출하고있습니다.

minosmlee commented on docs/data-sources/cdss_cluster.md:
````diff
@@ -0,0 +1,70 @@
+    storage_size      = [for k in data.ncloud_cdss_cluster.cluster.broker_nodes : k][0]["storage_size"]
+  }
+}
+```
````
datasource의 예시에는 filter를 이용하여 id를 찾을수있다 정도면 되기때문에 cluster-2의 예시는 없어도 될거같습니다.

minosmlee commented on docs/data-sources/cdss_config_group.md:
```diff
@@ -0,0 +1,23 @@
+In addition to all arguments above, the following attributes are exported
+
+* `name` - (Required) ConfigGroup name.
+* `description` - (Required) ConfigGroup description.
```
역시 filter (name, kafka_version_code) 기능을 추가해주시면 좋겠습니다.
config_group_uuid == id 일 경우 둘다 노출시키는 것이 좋을것 같습니다.

eastgerm committed:
Apply code review

wonchulee commented:
@eastgerm 님,
공공 베타 환경에서 수행시 아래와 같이 config_group 만 생성하더라도 500 에러가 발생하는 것 같은데, 혹시 한 번 확인 부탁드릴 수 있을까요?
```terraform
data "ncloud_cdss_kafka_version" "kafka-version" {
  filter {
    name   = "name"
    values = ["Kafka 2.8.2"]
  }
}

resource "ncloud_cdss_config_group" "config-group" {
  name = "from-tf-config"
  kafka_version_code = data.ncloud_cdss_kafka_version.kafka-version.id
  description = "test"
}
```

* 에러
2022-12-14T09:39:13.753+0900 [INFO]  provider.terraform-provider-ncloud: 2022/12/14 09:39:13 [ERROR] resourceNcloudCDSSConfigGroupCreate error params={"configGroupName":"from-tf-config","description":"test","kafkaVersionCode":"2823006"}, err=500 Internal Server Error: timestamp=2022-12-14T09:39:13.752+0900

wonchulee commented on ncloud/resource_ncloud_cdss_config_group.go:
```diff
@@ -0,0 +1,123 @@
+		KafkaVersionCode: *StringPtrOrNil(d.GetOk("kafka_version_code")),
+	}
+
+	logCommonRequest("resourceNcloudCDSSClusterCreate", reqParams)
```
사소한 부분이지만, CDSSConfigGroup이 아닌 CDSSCluster 이름으로 로깅되었네요!

eastgerm commented:
> @eastgerm 님, 공공 베타 환경에서 수행시 아래와 같이 config_group 만 생성하더라도 500 에러가 발생하는 것 같은데, 혹시 한 번 확인 부탁드릴 수 있을까요?
> 
> ```terraform
> data "ncloud_cdss_kafka_version" "kafka-version" {
>   filter {
>     name   = "name"
>     values = ["Kafka 2.8.2"]
>   }
> }
> 
> resource "ncloud_cdss_config_group" "config-group" {
>   name = "from-tf-config"
>   kafka_version_code = data.ncloud_cdss_kafka_version.kafka-version.id
>   description = "test"
> }
> ```
> 
> * 에러
>   2022-12-14T09:39:13.753+0900 [INFO]  provider.terraform-provider-ncloud: 2022/12/14 09:39:13 [ERROR] resourceNcloudCDSSConfigGroupCreate error params={"configGroupName":"from-tf-config","description":"test","kafkaVersionCode":"2823006"}, err=500 Internal Server Error: timestamp=2022-12-14T09:39:13.752+0900

```tf
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # ncloud_cdss_config_group.config-group will be created
  + resource "ncloud_cdss_config_group" "config-group" {
      + description        = "test"
      + id                 = (known after apply)
      + kafka_version_code = "2823006"
      + name               = "tf-config-1"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

ncloud_cdss_config_group.config-group: Creating...
ncloud_cdss_config_group.config-group: Creation complete after 1s [id=33]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

```
@wonchulee 님, 제 환경에선 정상 동작 하는듯한데요. 공공 베타 환경 소스코드 배포 버전이 잠시 어긋났던 것 같습니다. 다시 확인 부탁드립니다 !

eastgerm commented:
@wonchulee @minosmlee 님, 상세한 리뷰 감사합니다. 
위에 코멘트로 리뷰 해주신 내용들 반영하여 푸쉬 했습니다. 확인 부탁 드립니다^^

wonchulee commented on docs/data-sources/cdss_cluster.md:
```diff
@@ -0,0 +1,75 @@
+
+data "ncloud_cdss_cluster" "cluster"{
+  id = var.cdss_cluster_uuid
+}
```
@minosmlee 님이 리뷰해주신 내용으로 id 대신 필터를 사용해서 cluster를 찾을 수 있게 변경해주셨는데, example usage에서는 여전히 id를 사용하고 있는 부분이 남아 있는 것 같습니다. 요것도 함께 업데이트해주시면 좋을 것 같아요

wonchulee commented on docs/data-sources/cdss_config_group.md:
```diff
@@ -0,0 +1,28 @@
+
+data "ncloud_cdss_config_group" "config_group"{
+  id = var.cdss_config_group_uuid
+}
```
cdss_cluster와 마찬가지로, 여기서 id 대신 필터를 사용하는 것으로 변경해주셨는데 예제에는 id를 사용하는 부분이 남아있어 이 부분도 함께 업데이트해주시면 좋겠습니다.

eastgerm committed:
Change docs example for cdss

wonchulee committed:
fix: Update cdss cluster resource/datasource

wonchulee commented:
아래 몇가지 수정 커밋을 포함하여 함께 쌓아두었습니다. 혹시 의도와 다르게 변경된 부분이 있다면 알려주세요

1. endpoints 하위의 `TypeString` 타입으로 구성된 list를 `TypeList`로 변경 (resource, datasource 모두, `plaintext`, `tls`, `zookeeper` 3가지 항목은 `,`를 구분자로 사용하고 나머지 항목은 `\n`을 구분자로 사용, 혹시 잘못된 부분이 있다면 알려주세요)
2. nested structure로 구성하기 위해 `TypeList` 타입으로 정의된 `cmak`, `manager_node`, `broker_node` argument의 `MaxItems`를 `1`로 설정
3. cdss_cluster datasource 의 endpoints attribute를 `optional` -> `computed` 속성으로 변경 (리뷰 후 resource는 computed 로 변경되었으나, datasource 는 누락된 것으로 보임)

wonchulee commented:
마스터 노드의 count가 변경되면 `ForceNew`로 리소스가 재생성 됩니다.  콘솔에서는 리소스 감소는 불가능하지만 리소스를 증가시킬 수 있는 인터페이스를 제공하고 있는데요, 리소스 재생성 없이 마스터 노드를 증가시킬 수 있다면 (당장은 아니고 추후에라도) `ForceNew`를 제거하고, 노드를 증가시킬 수 있는 기능을 추가하는 것이 좋겠습니다. (리소스 감소는 예외처리)

wonchulee committed:
fix: Update uuid->id

eastgerm commented:
@wonchulee 님, 변경하신 의도와 내용 확인했고 문제 없을 듯 합니다. 신경 써주셔서 감사합니다

minosmlee commented on docs/data-sources/cdss_os_product.md:
```diff
@@ -0,0 +1,32 @@
+# Data Source: ncloud_cdss_node_os_image
```
문서 이름이 cdss_os_product -> cdss_node_os_image로 변경된것 같습니다.
하지만 아직 docs와 실제 datasource 코드 파일 이름이 cdss_os_product으로 남아있습니다.
이 data source의 성격 상 node_os_image 라는 이름이 더 적절한거 같긴합니다.

eastgerm replied:
확인 감사합니다. node_os_image.md로 파일명 수정하겠습니다.

minosmlee commented on docs/data-sources/cdss_os_product.md:
```diff
@@ -0,0 +1,32 @@
+## Attributes Reference
+
+* `id` - The ID of server image product.
+* `product_name` - Os image name
```
node_os_image 로 결정한다면 product_name을 image_name 으로 변경하는 것이 좋습니다.

eastgerm replied:
os_image로 결정했고, 관련된 resource,datasource,docs,example에도 변경 적용하겠습니다

minosmlee commented:
datasource : ncloud_cdss_kafka_version
datasource : ncloud_cdss_node_os_image (ncloud_cdss_os_product)
datasource : ncloud_cdss_node_product

에 대해 복수형의 datasource들을 추가하시면 좋겠습니다.
datasource : ncloud_cdss_kafka_versions
datasource : ncloud_cdss_node_os_images (ncloud_cdss_os_products)
datasource : ncloud_cdss_node_products

minosmlee commented on docs/data-sources/cdss_kafka_version.md:
```diff
@@ -0,0 +1,32 @@
+# Data Source: ncloud_cdss_kafka_versions
```
ncloud_cdss_kafka_versions -> ncloud_cdss_kafka_version
복수형 제목 -> 단수형 제목

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,136 @@
+
+  cmak {
+    user_name = "terraform"
+    user_password = var.cmak_user_password
```
user password 를 변경시 force replace가 발생합니다.
패스워드를 초기화 할 수 있는 메뉴가 존재하는 것같은데, 연계하여 modify in place로 동작하면 좋겠습니다.

minosmlee replied:
cmak_public_domain_enabled  옵션 추가를 할 수 있을까요?

minosmlee replied:
> user password 를 변경시 force replace가 발생합니다. 패스워드를 초기화 할 수 있는 메뉴가 존재하는 것같은데, 연계하여 modify in place로 동작하면 좋겠습니다.

잘되네요!

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,136 @@
+In addition to all arguments above, the following attributes are exported
+
+* `id` - Cluster id.
+* `endpoints` - .
+    * `plaintext` - List of broker nodes (Port 9092).
+    * `tls` - List of broker nodes (Port 9093).
+    * `public_endpoint_plaintext` - List of public endpoint of broker nodes.
+    * `public_endpoint_plaintext_listener_port` - List of listener port for public endpoint of broker nodes.
+    * `public_endpoint_tls` - List of public endpoint of broker nodes (TLS).
+    * `public_endpoint_tls_listener_port` - List of listener port for public endpoint of broker nodes (TLS).
+    * `hosts_private_endpoint_tls` - Editing details of the hosts file (Private IP hostname format).
+    * `hosts_public_endpoint_tls` - Editing details of the hosts file (Public IP hostname format).
+    * `zookeeper` - List of ZooKeeper nodes (Port 2181).
```
cluster 생성 후 아래와 같이 아웃풋에 "" 값들이 포함됩니다. 이 값들은 없애야 겠습니다.

``` hcl
  "endpoints" = [
    {
      "hosts_private_endpoint_tls" = [
        "10.0.1.19 cdss-tf-b-283f",
        "10.0.1.20 cdss-tf-b-283g",
        "10.0.1.21 cdss-tf-b-283h",
        "", <- 
      ]
      "hosts_public_endpoint_tls" = [
        "", <- 
      ]
      "plaintext" = [
        "10.0.1.19:9092",
        "10.0.1.20:9092",
        "10.0.1.21:9092",
      ]
      "public_endpoint_plaintext" = [
        "", <- 
      ]
      "public_endpoint_plaintext_listener_port" = [
        "cdss-tf-b-283f 9100",
        "cdss-tf-b-283g 9101",
        "cdss-tf-b-283h 9102",
        "", <- 
      ]
      "public_endpoint_tls" = [
        "", <- 
      ]
      "public_endpoint_tls_listener_port" = [
        "cdss-tf-b-283f 9300",
        "cdss-tf-b-283g 9301",
        "cdss-tf-b-283h 9302",
        "", <- 
      ]
      "tls" = [
        "cdss-tf-b-283f:9093",
        "cdss-tf-b-283g:9093",
        "cdss-tf-b-283h:9093",
      ]
      "zookeeper" = [
        "10.0.1.19:2181",
        "10.0.1.20:2181",
        "10.0.1.21:2181",
      ]
    },
  ]
```

wonchulee replied:
위 내용은 split을 사용하면서 발생한 부분으로 FieldsFunc을 사용하도록 변경하여 개선 업데이트 했습니다.

minosmlee replied:
확인하였습니다 감사합니다!

minosmlee commented:
> 마스터 노드의 count가 변경되면 `ForceNew`로 리소스가 재생성 됩니다. 콘솔에서는 리소스 감소는 불가능하지만 리소스를 증가시킬 수 있는 인터페이스를 제공하고 있는데요, 리소스 재생성 없이 마스터 노드를 증가시킬 수 있다면 (당장은 아니고 추후에라도) `ForceNew`를 제거하고, 노드를 증가시킬 수 있는 기능을 추가하는 것이 좋겠습니다. (리소스 감소는 예외처리)

@wonchulee 님 이건 broker node 를 의미하시는거죠? 

@eastgerm 님 broker node 증가는 꼭 modify in place 로 진행이 되어야될거같습니다. ValidateDiagFunc를 이용하여 3~10의 값을 validation check 하는 구문을 추가해주세요.

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  broker_nodes {
+    node_product_code = data.ncloud_cdss_node_product.node_sample.id
```
콘솔 메뉴와 api 상으로 force New 없이 노드 스펙을 변경할 수 있는 기능을 제공하는 것같습니다.
그렇다면 역시 테라폼에서도 modify in place로 진행이 되는것이 좋을것같습니다.

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  manager_node {
+    node_product_code = data.ncloud_cdss_node_product.node_sample.id
```
콘솔 메뉴와 api 상으로 force New 없이 노드 스펙을 변경할 수 있는 기능을 제공하는 것같습니다.
그렇다면 역시 테라폼에서도 modify in place로 진행이 되는것이 좋을것같습니다.

wonchulee committed:
fix(cdss): Filter empty item in the endpoint list

eastgerm committed:
fix(cdss): change name 'os_product' -> 'os_image'

eastgerm committed:
fix(cdss): change docs title 'cdss_kafka_versions' -> 'cdss_kafka_version'

eastgerm committed:
fix(cdss): add some of multiple datasource

- cdss_kafka_versions, cdss_node_products, cdss_os_images

eastgerm committed:
fix(cdss): apply number range validation checking for node_count

eastgerm committed:
feat(cdss): apply cmak password modify

eastgerm committed:
feat(cdss): apply node_count modify

eastgerm commented:
> cmak_public_domain_enabled 옵션 추가를 할 수 있을까요?

> 콘솔 메뉴와 api 상으로 force New 없이 노드 스펙을 변경할 수 있는 기능을 제공하는 것같습니다.
> 그렇다면 역시 테라폼에서도 modify in place로 진행이 되는것이 좋을것같습니다.

@minosmlee 님, 위 2가지 리뷰사항을 제외하고 나머지 부분 다 반영하여 푸쉬했습니다.

cmak_public_domain_enabled 옵션 같은 경우는 좀 더 플로우를 파악해야할듯 하여 아직 반영 못했고,
스펙 변경 같은 경우는 신규 api인데 go-sdk에 추가하지 못한 상태라 아직 반영 못했습니다.

위 두 사항은 완료되면 반영 후 공유 드리겠습니다.
우선은 나머지 반영 부분 먼저 확인 부탁드립니다!

eastgerm committed:
feat(cdss): apply node_product_code modify

eastgerm commented:
@minosmlee 님, 노드 스펙 변경 기능 적용하였습니다.

해당 go-sdk는 아직 릴리즈 전이라 아래 링크에서 클론 후 사용 부탁드립니다.
https://github.com/eastgerm/ncloud-sdk-go-v2/tree/vcdss_support_config_group

minosmlee commented on docs/data-sources/cdss_os_image.md:
````diff
@@ -0,0 +1,33 @@
+    values = ["CentOS 7.8 (64-bit)"]
+  }
+}
+```
````
```suggestion
# Data Source: ncloud_cdss_os_image

## Example Usage

```hcl
data "ncloud_cdss_os_image" "sample_01" {
  filter {
    name   = "id"
    values = ["SW.VCDSS.OS.LNX64.CNTOS.0708.B050"]
  }
}

data "ncloud_cdss_os_image" "sample_02" {
  filter {
    name   = "image_name"
    values = ["CentOS 7.8 (64-bit)"]
  }
}
```
```

minosmlee commented on docs/data-sources/cdss_os_images.md:
````diff
@@ -0,0 +1,13 @@
+
+```hcl
+data "ncloud_cdss_os_images" "os_images_sample" {}
+```
````
```suggestion
# Data Source: ncloud_cdss_os_images

## Example Usage

```hcl
data "ncloud_cdss_os_images" "os_images_sample" {}
```
```

minosmlee commented on docs/resources/cdss_cluster.md:
```diff
@@ -0,0 +1,136 @@
+    name = "product_name"
+    values = ["CentOS 7.8 (64-bit)"]
+  }
+}
```
```suggestion
data "ncloud_cdss_os_image" "os_sample" {
  filter {
    name = "image_name"
    values = ["CentOS 7.8 (64-bit)"]
  }
}
```

minosmlee commented:
> > 마스터 노드의 count가 변경되면 `ForceNew`로 리소스가 재생성 됩니다. 콘솔에서는 리소스 감소는 불가능하지만 리소스를 증가시킬 수 있는 인터페이스를 제공하고 있는데요, 리소스 재생성 없이 마스터 노드를 증가시킬 수 있다면 (당장은 아니고 추후에라도) `ForceNew`를 제거하고, 노드를 증가시킬 수 있는 기능을 추가하는 것이 좋겠습니다. (리소스 감소는 예외처리)
> 
> @wonchulee 님 이건 broker node 를 의미하시는거죠?
> 
> @eastgerm 님 broker node 증가는 꼭 modify in place 로 진행이 되어야될거같습니다. ValidateDiagFunc를 이용하여 3~10의 값을 validation check 하는 구문을 추가해주세요.

잘 되네요!

minosmlee commented on docs/resources/cdss_config_group.md:
````diff
@@ -0,0 +1,23 @@
+  description = "test"
+}
+```
+
````
``` hcl
ncloud_cdss_config_group.config_group: Creating...
╷
│ Error: Plugin did not respond
│ 
│   with ncloud_cdss_config_group.config_group,
│   on main-cdss.tf line 24, in resource "ncloud_cdss_config_group" "config_group":
│   24: resource "ncloud_cdss_config_group" "config_group" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-ncloud plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x101178c94]

goroutine 84 [running]:
github.com/terraform-providers/terraform-provider-ncloud/ncloud.resourceNcloudCDSSConfigGroupCreate({0x101564dc8, 0x1400007f4a0}, 0x0?, {0x1013c8ba0?, 0x1400007c940?})
        /Users/user/WORKS Drive/My Drive.localized/Coding/terraform/terraform-provider-ncloud/ncloud/resource_ncloud_cdss_config_group.go:60 +0x154
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x1400045aa80, {0x101564e00, 0x1400038f5c0}, 0xd?, {0x1013c8ba0, 0x1400007c940})
        /Users/user/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.13.0/helper/schema/resource.go:708 +0xe8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x1400045aa80, {0x101564e00, 0x1400038f5c0}, 0x140003db6c0, 0x1400010bf00, {0x1013c8ba0, 0x1400007c940})
        /Users/user/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.13.0/helper/schema/resource.go:838 +0x714
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x1400051ff38, {0x101564d58?, 0x1400025dd40?}, 0x140000afe00)
        /Users/user/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.13.0/helper/schema/grpc_provider.go:1021 +0xb70
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x14000494140, {0x101564e00?, 0x1400038e5d0?}, 0x1400033acb0)
        /Users/user/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.8.0/tfprotov5/tf5server/server.go:812 +0x384
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1015229e0?, 0x14000494140}, {0x101564e00, 0x1400038e5d0}, 0x1400007ed80, 0x0)
        /Users/user/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.8.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0x1400023c700, {0x1015686f8, 0x14000512340}, 0x1400093a5a0, 0x14000283c20, 0x101aab9e0, 0x0)
        /Users/user/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1282 +0xb28
google.golang.org/grpc.(*Server).handleStream(0x1400023c700, {0x1015686f8, 0x14000512340}, 0x1400093a5a0, 0x0)
        /Users/user/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1619 +0x82c
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /Users/user/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:921 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /Users/user/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:919 +0x290

Error: The terraform-provider-ncloud plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


민간 리얼에서는 config group이 생성이 되지 않는데, 확인이 필요할거같습니다.

eastgerm replied:
@minosmlee 님, config group의 description값에 널체크 추가 했고 description값 변경에 따른 modify in place 동작도 추가하여 반영했습니다.

minosmlee replied:
감사합니다. 이제 잘 되네요!

minosmlee committed:
Apply suggestions from code review

eastgerm committed:
feat(cdss): apply config group description modify & null check

eastgerm committed:
feat(cdss): apply cluster config_group_no modify

youngmn reviewed (approved):
LGTM 👍

wonchulee committed:
fix: Resolve function redeclared conflicts with SES

ncloud/resource_ncloud_ses_cluster.go:529:6: getChangedNodeProductCode redeclared in this block
	ncloud/resource_ncloud_cdss_cluster.go:468:6: other declaration of getChangedNodeProductCode
ncloud/resource_ncloud_ses_cluster.go:529:6: getChangedNodeProductCode redeclared in this block
	ncloud/resource_ncloud_cdss_cluster.go:468:6: other declaration of getChangedNodeProductCode

wonchulee commented:
https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/255 반영 후 컨플릭이 발생하는 부분 & function 재정의로 빌드 에러가 발생하는 부분 수정해서 리베이스 했습니다.

eastgerm commented:
### Product
Cloud Data Streaming Service(VPC)
### Features
- New Resource: `cdss_cluster` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Resource: `cdss_config_group` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_cluster` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_config_group` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_kafka_version` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_kafka_versions` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_node_product` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_node_products` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_os_image` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)
- New Data Source: `cdss_os_images` (https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/248)

jjanga303 commented:
민간/금융 리얼에서 QA완료되었습니다.
- 클러스터 생성
- 클러스터 삭제
- 클러스터 노드 스펙 변경
- 클러스터 노드 추가
- 클러스터 Config Group 변경
- 클러스터 패스워드 초기화
- Config Group 생성
- Config Group 편집
- Config Group 삭제
